docs/proposals/kubelet-tls-bootstrap.md

```diff
@@ -0,0 +1,231 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubelet TLS bootstrap
+
+Author: George Tankersley (george.tankersley@coreos.com)
+
+## Preface
+
+This document describes a method for a kubelet to bootstrap itself
+into a TLS-secured cluster. Crucially, it automates the provision and
+distribution of signed certificates.
+
+## Overview
+
+When a kubelet runs for the first time, it must be given TLS assets
+or generate them itself. In the first case, this is a burden on the cluster
+admin and a significant logistical barrier to secure Kubernetes rollouts. In
+the second, the kubelet must self-sign its certificate and forfeits many of the
+advantages of a PKI system. Instead, we propose that the kubelet generate a
+private key and a CSR for submission to a cluster-level certificate signing
+process.
```

Comment:
I don't want to scope creep this, but I would like a section that describes how other usecases for CSR on the cluster overlaps, and make sure this is not so narrowly scoped that it can't be useful elsewhere.

Reply:
Hi all - sorry to dig up an old thread!

I'm working on cert-manager, which is effectively 'kube-lego 2.0'. It aims to generalise obtaining certificates in Kubernetes, specifically for applications (which it looks like this comment thread touched on way back when).

We're looking at this moving into the kubernetes-incubator, and have a few open questions around existing work. I see the k8s certificates API, and we are finding ourselves effectively re-implementing it.

It seems like this document is the canonical design proposal for the Certificates API group itself, but I can't seem much information on the expected uses of the API.

As it stands, I can see a few issues with the existing certificates API that prevent us being able to adopt it:

* CSRs are non-namespaced - users are often scoped to only particular namespaces, but still need to make requests. There's no way for us to support this right now, and so users would need create/update/delete permission for all CSR resources, which breaks a lot of isolation models.

* CSRs directly embed signed certificates in their API status - this may not be too much of an issue, as the Certificate itself needn't be classed as 'secret'.

* Lack of ability to specify 'issuing source' - so if you look at the cert-manager proposal, we intend to support a variety of issuing sources. This includes a simple CA keypair (like the kube-controller-manager does right now), as well as others such as ACME, and in future potentially Vault. There is no way to 'choose' which issuer should issue a CSR as part of the certificates API, meaning users would have to go "all or nothing" with different issuers. In cert-manager, the `Certificate` resource (the equivalent of the `CertificateSigningRequest` resource) has an `issuer` field which allows us to disambiguate. We also allow extra configuration for the issuer to be specified (e.g. in the ACME case, which challenge mechanism should be used)

Reply:
For reference, I have a draft proposal drawn up for cert-manager already: https://github.com/jetstack-experimental/cert-manager/pull/50